mmpose/models/heads/base_head.py

```diff
@@ -29,3 +33,126 @@ def predict(self, feats: Tuple[Tensor], batch_data_samples: OptSampleList,
         """Calculate losses from a batch of inputs and data samples."""
+
+    def _get_in_channels(self):
+        """Get the input channel number of the network according to the feature
+        channel numbers and the input transform type."""
+
+        feat_channels = self.in_channels
+        if isinstance(feat_channels, int):
+            feat_channels = [feat_channels]
+
+        if self.input_transform == 'resize_concat':
+            if isinstance(self.input_index, int):
+                in_channels = feat_channels[self.input_index]
+            else:
+                in_channels = sum(feat_channels[i] for i in self.input_index)
+        elif self.input_transform == 'select':
+            if isinstance(self.input_index, int):
+                in_channels = feat_channels[self.input_index]
+            else:
+                in_channels = [feat_channels[i] for i in self.input_index]
+        else:
+            raise (ValueError,
+                   f'Invalid input transform mode "{self.input_transform}"')
+
+        return in_channels
+
+    def _transform_inputs(
+            self, feats: Tuple[Tensor]) -> Union[Tensor, Tuple[Tensor]]:
+        """Transform multi scale features into the network input."""
+        if self.input_transform == 'resize_concat':
+            inputs = [feats[i] for i in self.input_index]
+            resized_inputs = [
+                F.interpolate(
+                    x,
+                    size=inputs[0].shape[2:],
+                    mode='bilinear',
+                    align_corners=self.align_corners) for x in inputs
+            ]
+            inputs = torch.cat(resized_inputs, dim=1)
+        elif self.input_transform == 'select':
+            if isinstance(self.input_index, int):
+                inputs = feats[self.input_index]
+            else:
+                inputs = tuple(feats[i] for i in self.input_index)
+        else:
+            raise (ValueError,
+                   f'Invalid input transform mode "{self.input_transform}"')
+
+        return inputs
+
+    def decode(self, batch_outputs: Union[Tensor, Tuple[Tensor]],
+               batch_data_samples: OptSampleList,
+               test_cfg: ConfigType) -> SampleList:
+        """Decode keypoints from outputs."""
+
+        if self.decoder is None:
+            raise RuntimeError(
+                f'The decoder has not been set in {self.__class__.__name__}. '
+                'Please set the decoder configs in the init parameters to '
+                'enable head methods `head.predict()` and `head.decode()`')
+
+        if batch_data_samples is None:
+            raise ValueError(
+                '`batch_data_samples` is required to decode keypoitns.')
+
+        if isinstance(batch_outputs, Tensor):
+            batch_outputs_np = _to_numpy(batch_outputs)
+            batch_heatmaps_np = batch_outputs_np
+
+            head_type = type(self).__name__.lower()
+            if 'regression' in head_type:
```

Comment:
If we move 'decode()' to the base class, I think it's better to assign `pred_fields.heatmap` in `predict()` instead of `decode()` since it's more head-specific.

Reply:
Besides, it may be risky to control the behavior by the class name. This won't be a problem if we assign `pred_fields.heatmap` in subclasses.

Reply:
Good suggestion.